Python/pythonrun.c

```diff
@@ -1253,183 +1253,214 @@ print_chained(struct exception_print_context* ctx, PyObject *value,
+        return 0;
+    }
+
+    PyObject *cause = PyException_GetCause(value);
+    if (cause) {
+        int in_seen;
+        PyObject *check_id = PyLong_FromVoidPtr(cause);
+        if (check_id == NULL) {
+            in_seen = -1;
+        } else {
+            in_seen = PySet_Contains(ctx->seen, check_id);
+            Py_DECREF(check_id);
+        }
+        int err = 0;
+        if (in_seen == 0) {
+            err = print_chained(ctx, cause, cause_message, "cause");
+        }
+        else if (in_seen == -1) {
+            /* Ignore errors in 'seen' lookup */
             PyErr_Clear();
-        else if (PyExceptionInstance_Check(value)) {
-            PyObject *check_id = NULL;
-
-            cause = PyException_GetCause(value);
-            context = PyException_GetContext(value);
-            if (cause) {
-                check_id = PyLong_FromVoidPtr(cause);
-                if (check_id == NULL) {
-                    res = -1;
-                } else {
-                    res = PySet_Contains(ctx->seen, check_id);
-                    Py_DECREF(check_id);
-                }
-                if (res == -1)
-                    PyErr_Clear();
-                if (res == 0) {
-                    err = print_chained(ctx, cause, cause_message, "cause");
-                }
-            }
-            else if (context &&
-                !((PyBaseExceptionObject *)value)->suppress_context) {
-                check_id = PyLong_FromVoidPtr(context);
-                if (check_id == NULL) {
-                    res = -1;
-                } else {
-                    res = PySet_Contains(ctx->seen, check_id);
-                    Py_DECREF(check_id);
-                }
-                if (res == -1)
-                    PyErr_Clear();
-                if (res == 0) {
-                    err = print_chained(ctx, context, context_message, "context");
-                }
-            }
-            Py_XDECREF(context);
-            Py_XDECREF(cause);
         }
-        Py_XDECREF(value_id);
+        Py_DECREF(cause);
+        return err;
     }
-    if (err) {
-        /* don't do anything else */
+    if (((PyBaseExceptionObject *)value)->suppress_context) {
+        return 0;
     }
-    else if (!_PyBaseExceptionGroup_Check(value)) {
-        print_exception(ctx, value);
+    PyObject *context = PyException_GetContext(value);
+    if (context) {
+        int in_seen;
+        PyObject *check_id = PyLong_FromVoidPtr(context);
+        if (check_id == NULL) {
+            in_seen = -1;
+        } else {
+            in_seen = PySet_Contains(ctx->seen, check_id);
+            Py_DECREF(check_id);
+        }
+        int err = 0;
+        if (in_seen == 0) {
+            err = print_chained(ctx, context, context_message, "context");
+        }
+        else if (in_seen == -1) {
+            /* Ignore errors in 'seen' lookup */
+            PyErr_Clear();
+        }
+        Py_DECREF(context);
+        return err;
```

Comment:
Can you refactor out the "context" and the "cause" if blocks into a common function?